### PR TITLE
HEEDLS-899 Fix supervisor delegate relationships

### PR DESCRIPTION
**Pull request by livzorn**
### JIRA link
[_HEEDLS-899_](https://softwiretech.atlassian.net/browse/HEEDLS-899)

### Description
Ensured all registration journeys still update the supervisor delegate record as expected, and changed them to match the delegate email to either the primary or centre specific email.
Also updated SubmitSummary in SelfAssessment.cs to save the CentreSpecificEmail ?? PrimaryEmail as the delegate's email for a new SupervisorDelegate record, and updated GetUserEmail in the SupervisorController to get the supervisor's CentreSpecificEmail ?? PrimaryEmail.

-----
### Developer checks
(Leave tasks unticked if they haven't been appropriate for your ticket.)

I have:
- [x] Run the formatter and made sure there are no IDE errors.
- [x] Written tests for the changes (accessibility tests, unit tests for controller, data services, services, view models, etc)
- [x] Manually tested my work with and without JavaScript. Full manual testing guidelines can be found here: https://softwiretech.atlassian.net/wiki/spaces/HEE/pages/6703648740/Testing
- [ ] Updated/added documentation in Swiki and/or Readme. Links (if any) are below:
  - [doc_1_here](link_1_here)
- [ ] Updated my Jira ticket with information about other parts of the system that were touched as part of the MR and have to be sanity tested to ensure nothing’s broken.
- [x] Scanned over my own MR to ensure everything is as expected.

### CODE REVIEW

File: DigitalLearningSolutions.Data.Tests/DataServices/SupervisorDelegateDataServiceTests.cs [hunk 1]
```diff
@@ -0,0 +1,123 @@
+﻿namespace DigitalLearningSolutions.Data.Tests.DataServices
+{
+    using System;
+    using System.Collections.Generic;
+    using System.Linq;
+    using System.Transactions;
+    using Dapper;
+    using DigitalLearningSolutions.Data.DataServices;
+    using DigitalLearningSolutions.Data.Models.Email;
```
Comment (simonbrent) [inline DigitalLearningSolutions.Data.Tests/DataServices/SupervisorDelegateDataServiceTests.cs:9]: `DigitalLearningSolutions.Data.Models.Email` is unused

File: DigitalLearningSolutions.Data.Tests/DataServices/SupervisorDelegateDataServiceTests.cs [hunk 2]
```diff
@@ -0,0 +1,123 @@
...
+        [SetUp]
+        public void Setup()
+        {
+            connection = ServiceTestHelper.GetDatabaseConnection();
+            var logger = A.Fake<ILogger<SupervisorDelegateDataService>>();
+            supervisorDelegateDataService = new SupervisorDelegateDataService(connection, logger);
+        }
+
+        [Test]
+        public void GetSupervisorDelegateRecordByInviteHash_returns_correct_record()
```
Comment (simonbrent) [inline DigitalLearningSolutions.Data.Tests/DataServices/SupervisorDelegateDataServiceTests.cs:35]: The query in `GetSupervisorDelegateRecordByInviteHash` is wrapped in a try/catch, but there is only a test that it returns the correct record. There should also be a test that it logs an error and returns null if the same `inviteHash` exists for multiple rows

File: DigitalLearningSolutions.Data.Tests/DataServices/SupervisorDelegateDataServiceTests.cs [hunk 3]
```diff
@@ -0,0 +1,123 @@
...
+                {
+                    delegateEmailForValidRecord,
+                    delegateEmailForRecordWithNonNullCandidateId,
+                    delegateEmailForRemovedRecord,
+                    nullDelegateEmail,
+                }
+            ).ToList();
+
+            // Then
+            result.Count.Should().Be(1);
```
Comment (simonbrent) [inline DigitalLearningSolutions.Data.Tests/DataServices/SupervisorDelegateDataServiceTests.cs:97]: Multiple assertions should be wrapped in `using (new AssertionScope())`

File: DigitalLearningSolutions.Data.Tests/DataServices/SupervisorDelegateDataServiceTests.cs [hunk 4]
```diff
@@ -0,0 +1,123 @@
...
+            const int candidateId = 1;
+
+            // When
+            supervisorDelegateDataService.UpdateSupervisorDelegateRecordsCandidateId(
+                new List<int> { 6, 7, 8 },
+                1
+            );
+
+            // Then
+            var updatedRecord =
```
Comment (simonbrent) [inline DigitalLearningSolutions.Data.Tests/DataServices/SupervisorDelegateDataServiceTests.cs:116]: This isn't explicitly testing that the candidateId has been changed, just that it is 1 after the update method has been called...it could have been 1 beforehand too. To remedy this:
* `var oldRecord = supervisorDelegateDataService.GetSupervisorDelegateRecordByInviteHash(inviteHashForFirstSupervisorDelegateRecord)` before updating
* Test that `oldRecord.CandidateID` is not 1
* Test that `oldRecord.ID` is in the list [6,7,8] (otherwise you aren't explicitly proving that the record you're updating is the same as the record you're getting from the database after the update)

**Review by simonbrent — CHANGES_REQUESTED** [review]
`DigitalLearningSolutions.Data/Services/SupervisorService.cs` contains lots of functions which do SQL queries. These should be probably be moved to a data service.

File: DigitalLearningSolutions.Data/Services/RegistrationService.cs [hunk 5]
```diff
@@ -503,11 +500,13 @@ private void ValidateCentreEmail(int userId, string centreEmail)
             DelegateRegistrationModel delegateRegistrationModel
         )
         {
+            var delegateEmails = new List<string?>
+                { delegateRegistrationModel.PrimaryEmail, delegateRegistrationModel.CentreSpecificEmail };
+
```
Comment (VladNicol) [inline DigitalLearningSolutions.Data/Services/RegistrationService.cs:505]: Unless I missed something, it looks like this is the only place which can trigger a call to `SupervisorDelegateDataService.GetPendingSupervisorDelegateRecordsByEmailsAndCentre` with an empty email in the list. Related to Simon's comment about having nulls in the email list, maybe it makes sense to only add to the list if the emails are not null here?

**Review by VladNicol — APPROVED** [review]
No further comments, other than what Simon spotted.

**Comment by SteveJacksonSoft** [comment]
> `DigitalLearningSolutions.Data/Services/SupervisorService.cs` contains lots of functions which do SQL queries. These should be probably be moved to a data service.

I mean ideally yes... but this is code that's being managed by HEE which we rarely touch. I'm not convinced it's worth the effort / risk of trying to split the data layer from the service layer.

**Comment by SteveJacksonSoft** [comment]
Ideally we should wait until https://softwiretech.atlassian.net/browse/HEEDLS-932 is merged as that will change the update logic slightly, so we'll have to make changes afterwards.

File: DigitalLearningSolutions.Data/DataServices/SupervisorDelegateDataService.cs [hunk 6]
```diff
@@ -73,10 +79,10 @@ public IEnumerable<SupervisorDelegate> GetPendingSupervisorDelegateRecordsByEmai
                     FROM SupervisorDelegates sd
                     INNER JOIN AdminUsers au ON sd.SupervisorAdminID = au.AdminID
                     WHERE au.CentreID = @centreId
-                      AND sd.DelegateEmail = @email
+                      AND sd.DelegateEmail IN @emails
```
[the remark this answers is not in the document]
Comment (SteveJacksonSoft) [reply DigitalLearningSolutions.Data/DataServices/SupervisorDelegateDataService.cs:82]: I'm not sure how Dapper implements this - could you test it? I'd expect it to work fine if `emails` is empty though...
It's possible for multiple supervisors to have invited the same delegate, so this method should return an IEnumerable.

[the remark this answers is not in the document]
Comment (simonbrent) [reply DigitalLearningSolutions.Data/DataServices/SupervisorDelegateDataService.cs:82]: Ok, it looks like an empty list results in `AND sd.DelegateEmail IN (SELECT @emails WHERE 1 = 0)`, i.e. a safe but unnecessary query.

If the first item in `emails` is null, you get `System.NotSupportedException : The first item in a list-expansion cannot be null` from Dapper, so that is a problem. Looking at the usages, it should not be possible to create a list where the first entry is null, because the list is created with primary email first, then centre specific, but this is non-obvious and fluky. I therefore propose:

* Change `GetPendingSupervisorDelegateRecordsByEmailsAndCentre` so that `emails` are non-nullable strings
* Change `GetPendingSupervisorDelegateIdsMatchingDelegate` so that `delegateRegistrationModel.CentreSpecificEmail` is only added to `delegateEmails` if it is not null

File: DigitalLearningSolutions.Web/Controllers/SupervisorController/SupervisorController.cs [hunk 7]
```diff
@@ -12,28 +12,28 @@
...
         private readonly IConfigDataService configDataService;
-        private readonly ICentreRegistrationPromptsService centreRegistrationPromptsService;
-        private readonly IUserDataService userDataService;
+        private readonly IFrameworkNotificationService frameworkNotificationService;
         private readonly ILogger<SupervisorController> logger;
-        private readonly IConfiguration config;
         private readonly ISearchSortFilterPaginateService searchSortFilterPaginateService;
+        private readonly ISelfAssessmentService selfAssessmentService;
+        private readonly ISupervisorService supervisorService;
+        private readonly IUserDataService userDataService;
```
[the remark this answers is not in the document]
Comment (livzorn) [reply DigitalLearningSolutions.Web/Controllers/SupervisorController/SupervisorController.cs:17]: removed all that. Yeah not sure either but HEE wrote those files so I'll leave them as is

File: DigitalLearningSolutions.Web.Tests/Controllers/LearningPortal/LearningPortalControllerTests.cs [hunk 8]
```diff
@@ -74,9 +75,9 @@ public void SetUp()
...
                 searchSortFilterPaginateService
             );
-            controller.ControllerContext = new ControllerContext { HttpContext = new DefaultHttpContext { User = user } };
+            controller.ControllerContext = new ControllerContext
+                { HttpContext = new DefaultHttpContext { User = user } };
             controller = controller.WithMockTempData();
-
         }
     }
 }
```
[the remark this answers is not in the document]
Comment (livzorn) [reply DigitalLearningSolutions.Web.Tests/Controllers/LearningPortal/LearningPortalControllerTests.cs:83]: It looks like there's no tests at all for the method I made my small change to. Since HEE wrote the method I'm not sure if we want to spend the time writing tests for it. @SteveJacksonSoft do you have an opinion?

File: DigitalLearningSolutions.Data/DataServices/SupervisorDelegateDataService.cs (hunk 6 above)
[the remark this answers is not in the document]
Comment (livzorn) [reply DigitalLearningSolutions.Data/DataServices/SupervisorDelegateDataService.cs:82]: Ah I see, I figured it would be fine since the primary email will never be null so it would always be called with at least one not null value, but better to write it so we only give it non-null values. 

I've changed the service method to filter out null/empty values so we only have valid emails when we call the data service

File: DigitalLearningSolutions.Data.Tests/DataServices/SupervisorDelegateDataServiceTests.cs (hunk 2 above)
Comment (livzorn) [reply DigitalLearningSolutions.Data.Tests/DataServices/SupervisorDelegateDataServiceTests.cs:35]: Wrote another test that it returns null. I was having toruble checking that a call to the logger was made though, got this error: `The current proxy generator can not intercept the method Microsoft.Extensions.Logging.LoggerExtensions.LogError(Microsoft.Extensions.Logging.ILogger logger, System.String message, System.Object[] args) for the following reason:
    - Extension methods can not be intercepted since they're static.`

File: DigitalLearningSolutions.Data.Tests/DataServices/SupervisorDelegateDataServiceTests.cs (hunk 3 above)
Comment (livzorn) [reply DigitalLearningSolutions.Data.Tests/DataServices/SupervisorDelegateDataServiceTests.cs]: done :)

File: DigitalLearningSolutions.Data.Tests/DataServices/SupervisorDelegateDataServiceTests.cs (hunk 4 above)
Comment (livzorn) [reply DigitalLearningSolutions.Data.Tests/DataServices/SupervisorDelegateDataServiceTests.cs]: Thanks, made those changes

File: DigitalLearningSolutions.Data.Tests/DataServices/SupervisorDelegateDataServiceTests.cs (hunk 1 above)
Comment (livzorn) [reply DigitalLearningSolutions.Data.Tests/DataServices/SupervisorDelegateDataServiceTests.cs]: removed

File: DigitalLearningSolutions.Web/Controllers/SupervisorController/SupervisorController.cs [hunk 9]
```diff
@@ -60,16 +60,9 @@ private int GetAdminID()
...
-            var userEmail = User.GetUserPrimaryEmail();
-            if (userEmail == null)
-            {
-                return "";
-            }
-            else
-            {
-                return userEmail;
-            }
+            var adminId = GetAdminID();
```
[the remark this answers is not in the document]
Comment (livzorn) [reply DigitalLearningSolutions.Web/Controllers/SupervisorController/SupervisorController.cs]: done

File: DigitalLearningSolutions.Web/Controllers/LearningPortalController/SelfAssessment.cs [hunk 10]
```diff
@@ -3,6 +3,7 @@
     using System;
     using System.Collections.Generic;
     using System.Linq;
+    using DigitalLearningSolutions.Data.DataServices.UserDataService;
```
[the remark this answers is not in the document]
Comment (livzorn) [reply DigitalLearningSolutions.Web/Controllers/LearningPortalController/SelfAssessment.cs:6]: removed

File: DigitalLearningSolutions.Data.Tests/DataServices/SupervisorDelegateDataServiceTests.cs (hunk 2 above)
Comment (simonbrent) [reply DigitalLearningSolutions.Data.Tests/DataServices/SupervisorDelegateDataServiceTests.cs:35]: Ok

File: DigitalLearningSolutions.Data/Services/SupervisorDelegateService.cs [hunk 11]
```diff
@@ -30,17 +34,27 @@ ISupervisorDelegateDataService supervisorDelegateDataService
...
         }
 
-        public void AddDelegateIdToSupervisorDelegateRecords(IEnumerable<int> supervisorDelegateIds, int delegateId)
+        public IEnumerable<SupervisorDelegate> GetPendingSupervisorDelegateRecordsByEmailsAndCentre(
+            int centreId,
+            IEnumerable<string?> emails
+        )
         {
-            supervisorDelegateDataService.UpdateSupervisorDelegateRecordsCandidateId(supervisorDelegateIds, delegateId);
+            emails = emails.Where(e => !string.IsNullOrWhiteSpace(e)).ToList();
```
Comment (simonbrent) [inline DigitalLearningSolutions.Data/Services/SupervisorDelegateService.cs:47]: It's normally not considered good practice to mutate input parameters to a function, so I think I'd prefer this to be 
```
var nonEmptyEmails = emails.Where(e => !string.IsNullOrWhiteSpace(e)).ToList();
```

**Review by simonbrent — CHANGES_REQUESTED** [review]

File: DigitalLearningSolutions.Web/Controllers/LearningPortalController/SelfAssessment.cs (hunk 10 above)
[the remark this answers is not in the document]
Comment (livzorn) [reply DigitalLearningSolutions.Web/Controllers/LearningPortalController/SelfAssessment.cs]: Sorry that's weird. Officially gone now

**Comment by livzorn** [comment]
I'll tackle the further changes once HEEDLS-932 is merged, or I'll leave TODO comments for them if it's not merged before I go on holiday next week
